pkg/controller/aws/identity/iamrole.go

```diff
@@ -0,0 +1,171 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package identity
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	awsiam "github.com/aws/aws-sdk-go-v2/service/iam"
+
+	"github.com/crossplaneio/crossplane-runtime/pkg/resource"
+	v1alpha1 "github.com/crossplaneio/crossplane/aws/apis/identity/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/clients/aws/iam"
+	"github.com/crossplaneio/crossplane/pkg/controller/aws/utils"
+)
+
+const (
+	errObjectNotIAMRole = "The managed resource is not an IAMRole resource"
+)
+
+// IAMRoleController is the controller for IAMRole objects
+type IAMRoleController struct{}
+
+// SetupWithManager creates a new Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func (c *IAMRoleController) SetupWithManager(mgr ctrl.Manager) error {
+	r := resource.NewManagedReconciler(mgr,
+		resource.ManagedKind(v1alpha1.IAMRoleGroupVersionKind),
+		resource.WithExternalConnecter(&roleConnector{client: mgr.GetClient()}),
+		resource.WithManagedConnectionPublishers())
+
+	name := strings.ToLower(fmt.Sprintf("%s.%s", v1alpha1.IAMRoleKindAPIVersion, v1alpha1.Group))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.IAMRole{}).
+		Complete(r)
+}
+
+type roleConnector struct {
+	client      client.Client
+	newClientFn func(*aws.Config) (iam.RoleClient, error)
+}
+
+func (conn *roleConnector) Connect(ctx context.Context, mgd resource.Managed) (resource.ExternalClient, error) {
+	cr, ok := mgd.(*v1alpha1.IAMRole)
+	if !ok {
+		return nil, errors.New(errObjectNotIAMRole)
+	}
+
+	awsconfig, err := utils.RetrieveAwsConfigFromProvider(ctx, conn.client, cr.Spec.ProviderReference)
+	if err != nil {
+		return nil, err
+	}
+
+	newClientFn := iam.NewRoleClient
+	if conn.newClientFn != nil {
+		newClientFn = conn.newClientFn
+	}
+
+	c, err := newClientFn(awsconfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create new IAMRole client")
+	}
+	return &roleExternal{c}, nil
+}
+
+type roleExternal struct {
+	client iam.RoleClient
+}
+
+func (e *roleExternal) Observe(ctx context.Context, mgd resource.Managed) (resource.ExternalObservation, error) {
+	cr, ok := mgd.(*v1alpha1.IAMRole)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errObjectNotIAMRole)
+	}
+
+	req := e.client.GetRoleRequest(&awsiam.GetRoleInput{
+		RoleName: aws.String(cr.Spec.RoleName),
+	})
+	req.SetContext(ctx)
+
+	observed, err := req.Send()
+
+	if iam.IsErrorNotFound(err) {
+		return resource.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	if err != nil {
+		return resource.ExternalObservation{}, err
+	}
+
+	cr.Status = cr.BuildStatusFromObservation(observed)
```

Comment:
It just occurred to me that `Observe` methods should update the `Ready` condition to reflect the current state of the external resource, by setting `Available`, `Creating`, etc per the below example. This is the responsibility of the CRUD methods rather than the managed resource reconciler just in case we want some flexibility in the types of conditions we set for some hypothetical future resource.

```go
	i.Status.State = existing.GetState().String()
	switch i.Status.State {
	case v1alpha2.StateReady:
		i.Status.SetConditions(runtimev1alpha1.Available())
		resource.SetBindable(i)
	case v1alpha2.StateCreating:
		i.Status.SetConditions(runtimev1alpha1.Creating())
	case v1alpha2.StateDeleting:
		i.Status.SetConditions(runtimev1alpha1.Deleting())
	}
```